File: Leetcode0032.java
class Solution {
    public int longestValidParentheses(String s) {
        //idea: traverse from left and right to find the maximum valid parentheses length
        if (s == null ||s.isEmpty()){
            return 0;
        }
        int left1 = 0, right1 = 0, left2 = 0, right2 = 0, maxLen = 0;
        for(int i=0;i<s.length();i++){
            char currCharLeft = s.charAt(i);
            char currCharRight = s.charAt(s.length()-1-i);
            //first check from the left
            if (currCharLeft=='('){
                left1 ++;
            }
            else{
                right1 ++;
                //check if we need to update the maximum length
                if (left1 == right1){
                    maxLen = Math.max(maxLen, 2*left1);
                }
                //check if we need to reset left and right
                if (right1>left1){
                    left1 = 0;
                    right1 = 0;
                }
            }
            //check from the right
            if (currCharRight == ')'){
                right2 ++;
            }
            else{
                left2 ++;
                // check if we need to update the maxLen
                if (left2==right2){
                     maxLen = Math.max(maxLen, 2*left2);
                }
                // check if we need to reset left2 and right2
                if (left2>right2){
                    left2 = 0;
                    right2 = 0;
                }
            }
        }
        return maxLen;
    }
}

File: Leetcode0032.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        #idea: traverse both from left and  right to find the longest valid parenthethese length
        if not s:
            return 0
        left1, right1, left2, right2, maxLen = 0,0,0,0,0
        for i in range(len(s)):
            currCharLeft = s[i]
            currCharRight = s[len(s)-1-i]
            # check the maximum left length
            # print i, left1, right1, left2, right2
            if currCharLeft == '(':
                # we can always add it
                left1 += 1
            else:
                right1 += 1
                #check if we can update the maxLen
                if left1 == right1:
                    maxLen = max(maxLen, 2*left1)
                # check if we need to reset left and right
                if right1>left1:
                    left1 = 0
                    right1 = 0
            # check thge maximum right length
            if currCharRight == ')':
                # we can always add it
                right2 += 1
            else:
                left2 += 1
                #check if we can update the maxLen
                if right2 == left2:
                    maxLen = max(maxLen, 2*right2)
                # check if we need to reset left and right
                if left2>right2:
                    right2 = 0
                    left2 = 0
        return maxLen